### PR TITLE
Add AgentRejectAction across multiple modules

### PR DESCRIPTION
**Pull request by li-boxuan**
This commit introduces the AgentRejectAction class and integrates it across various modules and actions. It includes updates to READMEs, action definitions, and agent controllers to handle the new 'reject' action. This functionality will allow agents to properly signal task rejection.

===========BELOW COMMIT MESSAGE IS GENERATED BY OPENDEVIN=============

Yes, again I am too lazy to write a commit message by myself. So I asked OpenDevin to do me the favour:

```bash
WORKSPACE_MOUNT_PATH="`PWD`" SANDBOX_TYPE="exec" poetry run python opendevin/core/main.py -t "write a commit message" -c ManagerAgent -d ./
```

Why this PR is needed?
1. To give the user and/or delegator agent a chance to rethink about the task
2. To encourage agent developers to think about edge/failure cases. E.g. without the prompt change included in this PR, `CommitWriterAgent` never finishes when running in a non-git folder
3. To differentiate between success and failure (ofc). Without this PR, we don't know if `There is no diff in the staging area; no changes to commit.` output from `CommitWriterAgent` really means there's no change to commit, OR the output itself is literally the compiled commit message.

TODO in this PR:
- [x] ~~Update `backend_architecture.puml`~~ This should be a separate issue.

Future work:
- [ ] Delegator should be able to rethink about the task assignment
- [ ] Micro-agents should have enforcements for their inputs & outputs schema
- [ ] UI changes to differentiate between "Finish" and "Reject"
- [ ] Any other agent needs to use "Reject" action?

### CODE REVIEW

File: opendevin/events/action/agent.py [hunk 1]
```diff
@@ -105,6 +105,17 @@ def message(self) -> str:
         return "All done! What's next on the agenda?"
 
 
+@dataclass
+class AgentRejectAction(Action):
+    outputs: Dict = field(default_factory=dict)
```
Comment (rbren) [inline opendevin/events/action/agent.py:110]: I don't think this should have an arbitrary dict of outputs, the way `finish` does. All we want is a `reason`

**Review by rbren — CHANGES_REQUESTED** [review]

**Comment by rbren** [comment]
Very cool! Agree this will be useful

**Review by rbren — APPROVED** [review]

File: opendevin/events/action/agent.py (hunk 1 above)
Comment (li-boxuan) [reply opendevin/events/action/agent.py:70]: Will create a follow-up to enforce the schema for micro-agents. We could discuss if we want to continue this approach there.

File: agenthub/dummy_agent/agent.py [hunk 2]
```diff
@@ -123,6 +124,10 @@ def __init__(self, llm: LLM):
                 'action': AgentFinishAction(),
                 'observations': [],
             },
+            {
+                'action': AgentRejectAction(),
+                'observations': [],
+            },
```
Comment (SmartManoj) [inline agenthub/dummy_agent/agent.py:130]: The agent_controller will stop after the previous action `AgentFinishAction`. If this action is placed before, the same will happen. So, this actions is needed?

Comment (li-boxuan) [reply agenthub/dummy_agent/agent.py:130]: Probably not needed

**Pull request merged**